File: source/neuropod/internal/config_utils.cc
/* Copyright (c) 2020 The Neuropod Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include "config_utils.hh"

#include "neuropod/internal/error_utils.hh"
#include "neuropod/internal/neuropod_loader.hh"

#include <json/json.h>

#include <fstream>
#include <sstream>
#include <stdexcept>
#include <unordered_map>

#include <cpp-semver.hpp>

namespace neuropod
{

namespace
{

[[noreturn]] void throw_neuropod_config_error(const std::string &message)
{
    NEUROPOD_ERROR("Error loading neuropod config! Please check your config file. {}", message);
}

const std::unordered_map<std::string, TensorType> type_mapping = {
    {"float32", FLOAT_TENSOR},
    {"float64", DOUBLE_TENSOR},
    {"string", STRING_TENSOR},

    {"int8", INT8_TENSOR},
    {"int16", INT16_TENSOR},
    {"int32", INT32_TENSOR},
    {"int64", INT64_TENSOR},

    {"uint8", UINT8_TENSOR},
    {"uint16", UINT16_TENSOR},
    {"uint32", UINT32_TENSOR},
    {"uint64", UINT64_TENSOR},
};

// Convert a string data type to a TensorType
TensorType convert_to_tensor_type(const Json::Value &dtype)
{
    if (!dtype.isString())
    {
        throw_neuropod_config_error("'dtype' must be a string.");
    }

    const auto dt  = dtype.asString();
    auto       got = type_mapping.find(dt);
    if (got == type_mapping.end())
    {
        std::stringstream ss;
        ss << "The specified data type '";
        ss << dt;
        ss << "' is invalid.";
        throw_neuropod_config_error(ss.str());
    }
    else
    {
        return got->second;
    }
}

std::vector<Dimension> get_dims_from_json(const Json::Value &json_shape)
{
    // Make sure that the shape is an array
    if (json_shape.isArray())
    {
        // The dims to return
        std::vector<Dimension> out;
        for (const auto &item : json_shape)
        {
            // Get the number and do some validation
            // Each item in the array must be either null or a positive integer
            if (item.isInt() && item.asInt64() > 0)
            {
                out.emplace_back(item.asInt64());
            }
            else if (item.isNull())
            {
                // A dim of size -1 means we won't check the size of that dim
                out.emplace_back(-1);
            }
            else if (item.isString())
            {
                // It is a symbol
                out.emplace_back(item.asString());
            }
            else
            {
                throw_neuropod_config_error(
                    "All items in 'shape' must be either null, a string, or a positive integer.");
            }
        }

        return out;
    }

    throw_neuropod_config_error("'shape' must be an array. Please check your config file");
}

} // namespace

Dimension::Dimension(int64_t v) : value(v) {}
Dimension::Dimension(std::string s) : value(-2), symbol(std::move(s)) {}

bool Dimension::operator==(const Dimension &other) const
{
    if (value == other.value)
    {
        // If it's a symbol, make sure the symbol name matches
        return value == -2 ? symbol == other.symbol : true;
    }

    return false;
}

TensorSpec::TensorSpec(std::string n, std::vector<Dimension> d, const TensorType t)
    : name(std::move(n)), dims(std::move(d)), type(t)
{
}

std::unique_ptr<ModelConfig> load_model_config(const std::string &neuropod_path)
{
    auto loader = get_loader(neuropod_path);

    // Load the config file
    auto stream = loader->get_istream_for_file("config.json");
    if (!stream)
    {
        NEUROPOD_ERROR("Error loading config file for neuropod '{}'", neuropod_path);
    }

    return load_model_config(*stream);
}

std::unique_ptr<ModelConfig> load_model_config(std::istream &input_stream)
{
    // Parse it
    Json::CharReaderBuilder rbuilder;
    Json::Value             obj;

    std::string parse_err;
    bool        parsingSuccessful = Json::parseFromStream(rbuilder, input_stream, &obj, &parse_err);

    if (!parsingSuccessful)
    {
        throw_neuropod_config_error("Error parsing JSON: " + parse_err);
    }

    // Make sure that name and platform are strings
    if (!obj["name"].isString() || !obj["platform"].isString())
    {
        throw_neuropod_config_error("'name' and 'platform' must be strings.");
    }

    const std::string  name        = obj["name"].asString();
    const std::string  platform    = obj["platform"].asString();
    const Json::Value &input_spec  = obj["input_spec"];
    const Json::Value &output_spec = obj["output_spec"];

    // By default, any version is okay
    std::string platform_version_semver = "*";
    if (obj.isMember("platform_version_semver"))
    {
        platform_version_semver = obj["platform_version_semver"].asString();
    }

    // Make sure that it's a valid semver version or range
    if (!semver::valid(platform_version_semver))
    {
        throw_neuropod_config_error("The provided platform version (" + platform_version_semver +
                                    ") was not a valid semver version or range. See https://semver.org/ and "
                                    "https://docs.npmjs.com/misc/semver#ranges");
    }

    // Get the inputs
    std::vector<TensorSpec> inputs;
    for (const auto &spec : input_spec)
    {
        // Make sure name is a string
        if (!spec["name"].isString())
        {
            throw_neuropod_config_error("'name' must be a string.");
        }

        inputs.emplace_back(
            spec["name"].asString(), get_dims_from_json(spec["shape"]), convert_to_tensor_type(spec["dtype"]));
    }

    // Get the outputs
    std::vector<TensorSpec> outputs;
    for (const auto &spec : output_spec)
    {
        // Make sure name is a string
        if (!spec["name"].isString())
        {
            throw_neuropod_config_error("'name' must be a string.");
        }

        outputs.emplace_back(
            spec["name"].asString(), get_dims_from_json(spec["shape"]), convert_to_tensor_type(spec["dtype"]));
    }

    // Get the list of custom ops if any
    std::vector<std::string> custom_ops;
    if (obj.isMember("custom_ops"))
    {
        const Json::Value &items = obj["custom_ops"];
        for (const auto &item : items)
        {
            custom_ops.emplace_back(item.asString());
        }
    }

    // Load the device mapping if any
    std::unordered_map<std::string, NeuropodDeviceType> input_tensor_device;
    if (obj.isMember("input_tensor_device"))
    {
        const Json::Value &device_mapping = obj["input_tensor_device"];
        const auto         names          = device_mapping.getMemberNames();
        for (const auto &tensor_name : names)
        {
            const auto type = device_mapping[tensor_name].asString();
            if (type == "GPU")
            {
                input_tensor_device[tensor_name] = DeviceType::GPU;
            }
            else if (type == "CPU")
            {
                input_tensor_device[tensor_name] = DeviceType::CPU;
            }
            else
            {
                throw_neuropod_config_error("Invalid device type: " + type);
            }
        }
    }
    else
    {
        // Default all the tensors to GPU
        // TODO(vip): Remove this on version increase
        for (const auto &input : inputs)
        {
            input_tensor_device[input.name] = DeviceType::GPU;
        }
    }

    // Not directly using make_unique because of brace initialization
    return stdx::make_unique<ModelConfig>(
        ModelConfig{name, platform, platform_version_semver, inputs, outputs, custom_ops, input_tensor_device});
}

} // namespace neuropod

File: source/python/neuropod/utils/config_utils.py
# Copyright (c) 2020 UATC, LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
import os
from six import string_types, integer_types

from neuropod.utils.dtype_utils import get_dtype_name

ALLOWED_DTYPES = [
    "float32",
    "float64",
    "string",
    "int8",
    "int16",
    "int32",
    "int64",
    "uint8",
    "uint16",
    "uint32",
    "uint64",
]


def validate_tensor_spec(spec):
    """
    Validates a tensor spec
    """
    for item in spec:
        name = item["name"]
        dtype = item["dtype"]
        shape = item["shape"]

        if dtype not in ALLOWED_DTYPES:
            raise ValueError("{} is not an allowed data type!".format(dtype))

        if not isinstance(name, string_types):
            raise ValueError(
                "Field 'name' must be a string! Got value {} of type {}.".format(
                    name, type(name)
                )
            )

        if not isinstance(shape, (list, tuple)):
            raise ValueError(
                "Field 'shape' must be a tuple! Got value {} of type {}.".format(
                    shape, type(shape)
                )
            )

        for dim in shape:
            # A bool is an instance of an int so we have to do that check first
            is_uint = (
                (not isinstance(dim, bool))
                and isinstance(dim, integer_types)
                and dim > 0
            )

            if dim is None or is_uint or isinstance(dim, string_types):
                continue
            else:
                raise ValueError(
                    "All items in 'shape' must either be None, a string, or a positive integer! Got {}".format(
                        dim
                    )
                )


def validate_neuropod_config(config):
    """
    Validates a neuropod config
    """
    name = config["name"]
    platform = config["platform"]
    device_mapping = config["input_tensor_device"]

    if not isinstance(name, string_types):
        raise ValueError(
            "Field 'name' in config must be a string! Got value {} of type {}.".format(
                name, type(name)
            )
        )

    if not isinstance(platform, string_types):
        raise ValueError(
            "Field 'platform' in config must be a string! Got value {} of type {}.".format(
                platform, type(platform)
            )
        )

    validate_tensor_spec(config["input_spec"])
    validate_tensor_spec(config["output_spec"])

    # Optional custom ops
    if "custom_ops" in config:
        custom_ops = config["custom_ops"]

        if not isinstance(custom_ops, list):
            raise ValueError(
                "Optional field 'custom_ops' must be a list! Got value {} of type {}".format(
                    custom_ops, type(custom_ops)
                )
            )

        for op in custom_ops:
            if not isinstance(op, string_types):
                raise ValueError(
                    "All items in 'custom_ops' must be strings! Got value {} of type {}.".format(
                        op, type(op)
                    )
                )

    # Ensure all inputs have a device specified
    input_tensor_names = {item["name"] for item in config["input_spec"]}
    device_tensor_names = set(device_mapping.keys())
    inputs_without_device = input_tensor_names - device_tensor_names
    devices_without_input = device_tensor_names - input_tensor_names

    if len(inputs_without_device) != 0:
        raise ValueError(
            "Some input tensors do not have devices specified: {}".format(
                inputs_without_device
            )
        )

    if len(devices_without_input) != 0:
        raise ValueError(
            "Devices were specified for some tensors not in the `input_spec`: {}".format(
                devices_without_input
            )
        )

    for tensor_name, device in device_mapping.items():
        if device not in ["GPU", "CPU"]:
            raise ValueError(
                "Device must either be 'GPU' or 'CPU'! Got value '{}' for tensor named '{}'.".format(
                    device, tensor_name
                )
            )


def canonicalize_tensor_spec(spec):
    """
    Converts the datatypes in a tensor spec to canonical versions
    (e.g. converts double to float64)
    """
    transformed = []
    for item in spec:
        transformed.append(
            {
                "name": item["name"],
                "dtype": get_dtype_name(item["dtype"]),
                "shape": item["shape"],
            }
        )
    return transformed


def write_neuropod_config(
    neuropod_path,
    model_name,
    platform,
    input_spec,
    output_spec,
    platform_version_semver="*",
    custom_ops=None,
    input_tensor_device=None,
    default_input_tensor_device="GPU",
    **kwargs
):
    """
    Creates the neuropod config file

    :param  neuropod_path:  The path to a neuropod package
    :param  model_name:     The name of the model (e.g. "my_addition_model")
    :param  platform:       The model type (e.g. "python", "pytorch", "tensorflow", etc.)

    :param  platform_version_semver: The required platform version specified as semver range
                                     See https://semver.org/, https://docs.npmjs.com/misc/semver#ranges
                                     or https://docs.npmjs.com/misc/semver#advanced-range-syntax for
                                     examples and more info. Default is `*` (any version is okay)
                                     Ex: `1.13.1` or `> 1.13.1`

    :param  input_spec:     A list of dicts specifying the input to the model.
                            Ex: [{"name": "x", "dtype": "float32", "shape": (None, )}]

    :param  output_spec:    A list of dicts specifying the output of the model.
                            Ex: [{"name": "y", "dtype": "float32", "shape": (None, )}]

    :param  input_tensor_device:    A dict mapping input tensor names to the device
                                    that the model expects them to be on. This can
                                    either be `GPU` or `CPU`. Any tensors in `input_spec`
                                    not specified in this mapping will use the
                                    `default_input_tensor_device` specified below.

                                    If a GPU is selected at inference time, Neuropod
                                    will move tensors to the appropriate devices before
                                    running the model. Otherwise, it will attempt to run
                                    the model on CPU and move all tensors (and the model)
                                    to CPU.

                                    See the docstring for `load_neuropod` for more info.

                                    Ex: `{"x": "GPU"}`

    :param  default_input_tensor_device:    The default device that input tensors are expected
                                            to be on. This can either be `GPU` or `CPU`.

    """
    if custom_ops is None:
        custom_ops = []

    if input_tensor_device is None:
        input_tensor_device = {}

    # Canonicalize the specs
    input_spec = canonicalize_tensor_spec(input_spec)
    output_spec = canonicalize_tensor_spec(output_spec)

    # Set up the device mapping
    device_mapping = {}
    for item in input_spec:
        name = item["name"]
        if name in input_tensor_device:
            # Use the device specified by the user
            device_mapping[name] = input_tensor_device[name]
        else:
            # Use the default device
            device_mapping[name] = default_input_tensor_device

    # TODO: Switch to prototext
    with open(os.path.join(neuropod_path, "config.json"), "w") as config_file:
        config = {
            "name": model_name,
            "platform": platform,
            "platform_version_semver": platform_version_semver,
            "input_spec": input_spec,
            "output_spec": output_spec,
            "custom_ops": custom_ops,
            "input_tensor_device": device_mapping,
        }

        # Verify that the config is correct
        validate_neuropod_config(config)

        # Write out the config as JSON
        json.dump(config, config_file, indent=4)


def read_neuropod_config(neuropod_path):
    """
    Reads a neuropod config

    :param  neuropod_path:  The path to a neuropod package
    """
    with open(os.path.join(neuropod_path, "config.json"), "r") as config_file:
        config = json.load(config_file)

        # For backwards compatibility
        # TODO(vip): Remove this on version increase
        if "input_tensor_device" not in config:
            # If there is no mapping in the configuration, move all tensors to
            # GPU by default
            config["input_tensor_device"] = {
                item["name"]: "GPU" for item in config["input_spec"]
            }

        # Verify that the config is correct
        validate_neuropod_config(config)

        return config
